src/h5web/vis-packs/core/heatmap/HeatmapVis.tsx

```diff
@@ -95,7 +93,29 @@ function HeatmapVis(props: Props) {
         }}
       >
         <PanZoomMesh />
-        <TooltipMesh {...tooltipFormatters} guides="both" />
+        <TooltipMesh
+          guides="both"
+          renderTooltip={(x, y) => {
+            const xi = abscissaToIndex(x);
+            const yi = ordinateToIndex(y);
+            const abscissa = abscissas[xi];
+            const ordinate = ordinates[yi];
+
+            if (renderTooltip) {
+              return renderTooltip({ abscissa, ordinate, xi, yi, x, y });
+            }
+
+            return (
+              <>
+                {`${abscissaLabel || 'x'}=${formatTooltipVal(abscissa)}, `}
+                {`${ordinateLabel || 'y'}=${formatTooltipVal(ordinate)}`}
+                <div className={styles.tooltipValue}>
+                  {formatTooltipVal(dataArray.get(yi, xi))}
+                </div>
+              </>
+            );
+          }}
+        />
```

Comment:
`TooltipMesh` now also accepts a `renderTooltip` function, but with only `x` and `y` as arguments. It is up to the visualization component to deal with these coordinates and render the content of the tooltip.

This is the simplest solution that we discussed. I really wasn't a fan of passing the `abscissas` and `ordinates` arrays (plus the `switchAtMidPoints` boolean) to `TooltipMesh` so it could compute the correct `abscissa` and `ordinate` values.

One reason is that it would have forced potential users of `TooltipMesh` to always generate `abscissas` and `ordinates` arrays, even when they'd be happy just doing `Math.round()` on the tooltip's raw coordinates. It would also have complexified the code and types of `TooltipMesh`, as we would have had to check whether an `ordinates` array was passed before creating the corresponding `ordinateToIndex` scale, and we would have had to enforce that `renderTooltip` would only receive `abscissa` and not `ordinate` when the `ordinates` array was not passed.

With this implementation, `TooltipMesh` remains completely uncoupled with the custom axis values feature.

Sure, it's a fair amount of code to add to `HeatmapVis` and `LineVis`... but I actually like the explicitness of it. I could have turned `useTooltipFormatters` into `useTooltipRenderer`, or moved the whole `<TooltipMesh ... />` block into a `HeatmapTooltipMesh` component, but it didn't feel necessary.

Let me know what you think.